controller/agent.go

```diff
@@ -51,3 +81,92 @@ func (self *Controller) agentOpSnapshotDb(c *bufio.ReadWriter) error {
 	_, err := c.WriteString("success\n")
 	return err
 }
+
+func (self *Controller) HandleCustomAgentAsyncOp(conn net.Conn) error {
+	logrus.Debug("received agent operation request")
+
+	appIdBuf := []byte{0}
+	_, err := io.ReadFull(conn, appIdBuf)
+	if err != nil {
+		return err
+	}
+	appId := appIdBuf[0]
+
+	if appId != AgentAppId {
+		logrus.WithField("appId", appId).Debug("invalid app id on agent request")
+		return errors.New("invalid operation for controller")
+	}
+
+	options := channel.DefaultOptions()
+	options.ConnectTimeout = time.Second
+	listener := channel.NewExistingConnListener(self.config.Id, conn, nil)
+	_, err = channel.NewChannel("agent", listener, channel.BindHandlerF(self.bindAgentChannel), options)
+	return err
+}
+
+func (self *Controller) agentOpAsyncSnapshotDb(m *channel.Message, ch channel.Channel) {
+	log := pfxlog.Logger()
+	if err := self.network.SnapshotDatabase(); err != nil {
+		log.WithError(err).Error("failed to snapshot db")
+		handler_common.SendOpResult(m, ch, "db.snapshot", err.Error(), false)
+	} else {
+		handler_common.SendOpResult(m, ch, "db.snapshot", "success", true)
+	}
+}
+
+func (self *Controller) agentOpRaftList(m *channel.Message, ch channel.Channel) {
+	members, err := self.raftController.ListMembers()
+	if err != nil {
+		handler_common.SendOpResult(m, ch, "raft.list", err.Error(), false)
+	}
+
+	result := &mgmt_pb.RaftMemberListResponse{}
+	for _, member := range members {
+		result.Members = append(result.Members, &mgmt_pb.RaftMember{
+			Id:       member.Id,
+			Addr:     member.Addr,
+			IsVoter:  member.Voter,
+			IsLeader: member.Leader,
+		})
+	}
+
+	if err = protobufs.MarshalTyped(result).ReplyTo(m).WithTimeout(time.Second).Send(ch); err != nil {
+		pfxlog.Logger().WithError(err).Error("failure sending raft member list response")
+	}
+}
+
+func (self *Controller) agentOpRaftJoin(m *channel.Message, ch channel.Channel) {
+	addr, found := m.GetStringHeader(AgentAddrHeader)
+	if !found {
+		handler_common.SendOpResult(m, ch, "raft.join", "address not supplied", false)
+		return
+	}
+
+	id, found := m.GetStringHeader(AgentIdHeader)
+	if !found {
+		handler_common.SendOpResult(m, ch, "raft.join", "id not supplied", false)
+		return
+	}
+
+	isVoter, found := m.GetBoolHeader(AgentIsVoterHeader)
+	if !found {
+		isVoter = true
+	}
+
+	if err := self.raftController.Join(raft2.ServerAddress(addr), raft2.ServerID(id), !isVoter); err != nil {
+		handler_common.SendOpResult(m, ch, "raft.join", err.Error(), false)
+		return
+	}
+	handler_common.SendOpResult(m, ch, "raft.join", "success", true)
+}
+
+func (self *Controller) agentOpRaftRemove(m *channel.Message, ch channel.Channel) {
+	//id := string(m.Body)
+
+	// TODO: make this work like Join where we test if we're bootstrapped yet
+	//if err := self.raftController.HandleRemove(); err != nil {
+	//	return err
+	//}
+	// _, err := c.WriteString("success\n")
+	handler_common.SendOpResult(m, ch, "raft.remove", "no yet implemented", false)
```

Comment:
todo?

Reply:
Unless you think it's a bad idea, I'm not going to focus on having the raft management pieces fully baked at this stage. I want this pass focused on having the command infrastructure solid, then will do another pass (or passes, more likely) focused on solidifying the raft management pieces.